Stories codelab edits (#3174)
GoogleChrome/web.dev: src/site/content/en/blog/codelab-stories-component/index.md

---
layout: codelab
title: Stories component codelab
authors:
  - adamargyle
description: |
  Learn how to build an experience similar to Instagram Stories on the web.
date: 2020-06-08
hero: hero.jpg
glitch: gui-challenges-stories
glitch_path: app/index.html
related_post: stories-component
tags:
  - css
  - dom
  - javascript
  - layout
  - mobile
  - ux
---

This codelab teaches you how to build an experience like Instagram Stories
on the web. We'll build the component as we go, starting with HTML, then CSS,
then JavaScript.

Check out my blog post [Building a Stories component on the web](/building-a-stories-component)

> to learn about the tradeoffs I had to make while building this component.

Reviewer: ```suggestion
to learn about the progressive enhancements made while building this component.
```

{% YouTube '5_sfnQDr1-o' %}

{% Aside %}
This post assumes that you're familiar with the Instagram Stories UX
and will use terminology from that experience (e.g. a "friend's story").
{% endAside %}

## Setup

{% Instruction 'remix', 'ol' %}
1. Open `app/index.html`.

## HTML

I always aim to use [semantic HTML](https://en.wikipedia.org/wiki/Semantic_HTML).
Since each friend can have any number of stories, I thought it was meaningful to use a
`<section>` element for each friend and an `<article>` element for each story.
Let's start from the beginning though. First, we need a container for our
stories component.

Add a `<div>` element to your `<body>`:

> ```html/0,2

Reviewer: i searched the handbook for this syntax, is it there!? cuz it's rad, ty

Author: Looks like we're missing it. I created #3229 to document it.

}
```

The `.seen` class will be added to a story that needs an exit.
I got the custom easing function (`cubic-bezier(0.4, 0.0, 1,1)`)
from Material Design's [Easing](https://material.io/design/motion/speed.html#easing)
guide (see the *Accerlerated easing* section).

{% Aside %}

> TODO(kayce): Explain the ampersand syntax

Reviewer: we can push to the spec draft? https://drafts.csswg.org/css-nesting-1/#nest-selector

Author: Perfect

{% endAside %}

If you've got a keen eye you probably noticed the `pointer-events: none`
declaration and are scratching your head right now. I'd say this is the only
downside of the solution so far. We need this because a `.seen.story` element
will be on top and will receive taps, even though it's invisible. By setting the

> `pointer-events` to `none`, we turn the glass story into a window, and steal

Reviewer: ```suggestion
`pointer-events` to `none`, we turn the glass story into a window, and steal no
```

more user interactions. Not too bad of a trade off, not too hard to manage here
in our CSS right now. We're not juggling `z-index`. I'm feeling good about this
still.

> TODO(kayce): We need to rephrase the last paragraph. Turning the glass story into
> a window? Stealing more user interactions? 

Reviewer: ```suggestion
```

I think the edit I made above solves this problem, thoughts? Sorry you had to inherit that confusing typo!

Author: Makes sense now

</figure>

### Setup

To start out, let's compute and store as much information as we can.
Add the following code to `app/js/index.js`:

```js
const stories = document.querySelector('.stories')
const median = stories.offsetLeft + (stories.clientWidth / 2)
```

Our first line of JavaScript grabs and stores a reference to our primary HTML
element root. The next line calculates where the middle of our element is, so we
can decide if a tap is to go forward or backward.

### State

Next we make a small object with some state relevant to our logic. In this
case, we're only interested in the current story. In our HTML markup, we can

> access it by grabbing the 1st user and their 1st story. Add the highlighted code

Reviewer: ```suggestion
access it by grabbing the 1st friend and their most recent story. Add the highlighted code
```

}
```

> <!-- TODO(kayce): All JS code needs semicolons: https://google.github.io/styleguide/jsguide.html#formatting-semicolons-are-required -->

Reviewer: :shudders: lol, jk, sgtm

Reviewer: oh yeah, my style is "what don't i have to write".. there's going to be **more than semicolons** needing added to my js  😂

Author: That's the only big contentious one that I can see because of the risk of introducing bugs if the implicit semicolon insertion doesn't work as expected

})
```

> If a click happens and it's not on a `<article>` element we bail and do nothing.

Reviewer: thanks for keeping my tone in here ❤️
